ops/testing.py

```diff
@@ -342,17 +359,31 @@ def __init__(self, unit_name):
         self._unit_status = None
 
     def relation_ids(self, relation_name):
-        return self._relation_ids_map[relation_name]
+        self._calls.append(('relation_ids', relation_name))
```

Comment:
It would be awesome if we could record the call for each method without having to copy this line for eeeeeeeevery method.

What about using a class decorator? is this too much magic? http://linkode.org/#TxUe0WImzhIRbORBemRlL

Reply:
I imagine you could do something like this with `__getattribute__` tricks as well. (though there you have to watch out for members vs methods.)

The one bit I don't particularly prefer is that it means args vs kwargs ends up in every method call, even where we wouldn't care whether it would be passed by name or passed by position. Which also leads to double nesting of tuples, etc.

So instead of being able to do:
```
        self.assertEqual([
            ('relation_ids', 'db'),
            ('relation_list', rel_id),
        ], harness.get_backend_calls(reset=True))
```

You end up with:
```
        self.assertEqual([
            ('relation_ids', ('db',), {}),
            ('relation_list', (rel_id,), {}),
        ], harness.get_backend_calls(reset=True))
```

Which has quite a bit more 'noise'. And suddenly that starts breaking if someone does:
```
- self.backend.relation_ids(name)
+ self.backend.relation_ids(relation_name=name)
```

Even though that isn't a functional change to what is being requested.

We could handle the visual noise with something like:
```
def decorator(orig_method):
    def wrapped(self, *args, **kwargs):
        full_args = (orig_method.__name__,) + args
        if kwargs:
            full_args = full_args + (kwargs,)
	self.calls.append(full_args)
	return orig_method(self, *args, **kwargs)
    return wrapped
```

Which in your test example gives:
```
$ py ./record_calls.py
<function record_calls.<locals>.decorator.<locals>.wrapped at 0x7ff2bac31d90>
<function record_calls.<locals>.decorator.<locals>.wrapped at 0x7ff2bac316a8>
12
123 456
[('foo', 12), ('bar', 123, {'b': 456})]
```

I'm happy enough with that, that I'll implement it. If you prefer something else we can discuss. (I'll live with the not-actually-functionally-different for now.)